File: 프로그래머스/3/42893. 매칭 점수/매칭 점수.py
import re
from collections import defaultdict


def solution(word, pages):
    answer = 0
    max_score = 0
    
    indexes = defaultdict(int)
    basic_scores = defaultdict(int)
    total_scores = defaultdict(int)
    external_urls = defaultdict(set)
    internal_urls = defaultdict(set)
    
    for index, page in enumerate(pages):
        word, page = word.lower(), page.lower()
        
        url = re.search(r'(<meta property.+content=")(https://\S*)"/>', page).group(2)
        basic_scores[url] += sum(x == word for x in re.findall('[a-z]+', page))
        indexes[url] = index
        
        for ex_url in re.findall(r'<a href="(https://\S*)">', page):
            external_urls[url].add(ex_url)
            internal_urls[ex_url].add(url)
        
    for url in basic_scores:
        link_score = basic_scores[url] + sum(basic_scores[target] / len(external_urls[target]) for target in internal_urls[url])
        
        if max_score < link_score:
            max_score = link_score
            answer = indexes[url]

    return answer

File: 프로그래머스/3/42893. 매칭 점수/매칭 점수.java
import java.util.*;
import java.util.regex.*;

class Solution {

    private static final Pattern URL_PATTERN  = Pattern.compile("<meta property[^>]+content=\"(https://[^\"]+)\"");
    private static final Pattern WORD_PATTERN = Pattern.compile("[a-z]+");
    private static final Pattern LINK_PATTERN = Pattern.compile("<a href=\"(https://[^\"]+)\"");

    public int solution(String word, String[] pages) {
        word = word.toLowerCase();

        int n = pages.length;
        String[] urls       = new String[n];
        int[]    scores     = new int[n];
        int[]    extCounts  = new int[n];
        Map<String, Integer> urlIndex = new HashMap<>();
        Map<String, List<Integer>> inbound = new HashMap<>();

        for (int i = 0; i < n; i++) {
            String page = pages[i].toLowerCase();

            Matcher um = URL_PATTERN.matcher(page);
            if (!um.find()) continue;
            urls[i] = um.group(1);
            urlIndex.put(urls[i], i);

            Matcher wm = WORD_PATTERN.matcher(page);
            while (wm.find()) if (wm.group().equals(word)) scores[i]++;

            Matcher lm = LINK_PATTERN.matcher(page);
            while (lm.find()) {
                extCounts[i]++;
                inbound.computeIfAbsent(lm.group(1), k -> new ArrayList<>()).add(i);
            }
        }

        int answer = 0;
        double maxScore = -1;

        for (int i = 0; i < n; i++) {
            if (urls[i] == null) continue;
            double score = scores[i];
            for (int from : inbound.getOrDefault(urls[i], Collections.emptyList())) {
                if (extCounts[from] > 0) score += (double) scores[from] / extCounts[from];
            }
            if (score > maxScore) {
                maxScore = score;
                answer = i;
            }
        }

        return answer;
    }
}